Memory leak on nightly checks

We are seeing some random OOMs when running our nightly extra tests on Mac runners e.g.,
```
 INFO ide_ci::program::command: sbt ℹ️ [info] Test run org.enso.interpreter.test.EqualsTest finished: 0 failed, 0 ignored, 19 total, 3.341s
 INFO ide_ci::program::command: sbt ℹ️ java.lang.OutOfMemoryError: Java heap space
 INFO ide_ci::program::command: sbt ℹ️ Dumping heap to <redacted>/enso/engine/runtime-integration-tests/target/scala-2.13/java_pid49785.hprof ...
 INFO ide_ci::program::command: sbt ℹ️ Heap dump file created [4323755974 bytes in 6.489 secs]
 INFO ide_ci::program::command: sbt ℹ️ [error] Test org.enso.interpreter.test.VectorSortTest failed: java.lang.AssertionError: EnsoContext can be GCed when context is closed ref still alive: org.enso.interpreter.runtime.EnsoContext@47d1bf07, took 0.0 sec
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.test.utils.MemoryUtils.assertReference(MemoryUtils.java:43)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.test.utils.MemoryUtils.assertGC(MemoryUtils.java:35)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.test.utils.ContextUtils.close(ContextUtils.java:149)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.test.utils.ContextUtils$CustomStatement.evaluate(ContextUtils.java:560)
 INFO ide_ci::program::command: sbt ℹ️ [error]     ...
```

Thanks to https://github.com/enso-org/enso/pull/14251 we now have [heap dumps](https://github.com/enso-org/enso/actions/runs/19285357400) in summaries so can investigate further where do we have leaks.
It's possible leaks not only happen on MacOS but also on other runners but Mac's are a bit underpowered which in this case is to our advantage.

--- a/lib/java/test-utils/src/main/java/org/enso/test/utils/MemoryUtils.java
+++ b/lib/java/test-utils/src/main/java/org/enso/test/utils/MemoryUtils.java
@@ -2,6 +2,8 @@ package org.enso.test.utils;
 
 import static org.junit.Assert.fail;
 
+import java.io.PrintWriter;
+import java.io.StringWriter;
 import java.lang.ref.Reference;
 import java.util.ArrayList;
 import org.graalvm.polyglot.Context;
@@ -11,18 +13,23 @@ final class MemoryUtils {
   private MemoryUtils() {}
 
   static void assertGC(String msg, boolean expectGC, Reference<?> ref) {
+    var buffer = new StringWriter();
+    var log = new PrintWriter(buffer);
     var memory = expectGC ? new ArrayList<>() : null;
     var retry = 3;
     for (var i = 1L; ; i *= 2) {
       try {
         var size = (int) Math.min(i, Integer.MAX_VALUE / 2);
         if (i >= 64) {
-          tryHarderToGc();
+          tryHarderToGc(log);
+          flushLog(buffer);
         }
-        if (checkAndAlloc(ref, memory, size)) {
+        if (checkAndAlloc(log, ref, memory, size)) {
           break;
         }
       } catch (OutOfMemoryError err) {
+        err.printStackTrace(log);
+        flushLog(buffer);
         // launch the JVM with
         //   -XX:+HeapDumpOnOutOfMemoryError
         //   -XX:HeapDumpPath=/tmp
@@ -32,22 +39,21 @@ final class MemoryUtils {
         }
       }
     }
-    assertReference(ref, expectGC, msg, memory);
+    assertReference(buffer, ref, expectGC, msg, memory);
   }
 
   private static void assertReference(
-      Reference<?> ref, boolean expectGC, String msg, ArrayList<Object> memory) {
+      Object log, Reference<?> ref, boolean expectGC, String msg, ArrayList<Object> memory) {
     var obj = ref.get();
     if (expectGC) {
       if (obj != null) {
-        fail(msg + " ref still alive: " + obj);
+        fail(msg + " ref still alive: " + obj + "\n" + log);
       }
     } else {
       if (obj == null) {
-        fail(msg + " ref has been cleaned: " + obj);
+        fail(msg + " ref has been cleaned: " + obj + "\n" + log);
       }
     }
-    System.getLogger("assertGC").log(System.Logger.Level.TRACE, "Cannot GC {0}", memory);
   }
 
   /**
@@ -59,20 +65,53 @@ final class MemoryUtils {
    * @param toAllocate
    * @return
    */
-  private static boolean checkAndAlloc(Reference<?> ref, ArrayList<Object> memory, int toAllocate) {
+  private static boolean checkAndAlloc(
+      PrintWriter log, Reference<?> ref, ArrayList<Object> memory, int toAllocate) {
     if (ref.get() == null) {
       return true;
     }
+    log.println("System.gc()");
     System.gc();
     if (memory != null) {
+      log.println("Allocating " + toAllocate + " bytes");
       memory.add(new byte[toAllocate]);
     }
     return false;
   }
 
-  private static void tryHarderToGc() {
-    try (var ctx = Context.create()) {
-      System.getLogger("assertGC").log(System.Logger.Level.TRACE, "Creating and closing {0}", ctx);
+  private static void tryHarderToGc(PrintWriter log) {
+    for (var t : Thread.getAllStackTraces().entrySet()) {
+      var threadName = t.getKey().getName();
+      if (threadName.contains("TruffleCompilerThread")) {
+        var foundCompilation = false;
+        for (var frame : t.getValue()) {
+          if ("com.oracle.truffle.runtime.CompilationTask".equals(frame.getClassName())) {
+            foundCompilation = true;
+          }
+        }
+        if (foundCompilation) {
+          log.println("Found running compiler thread: " + threadName);
+          for (var frame : t.getValue()) {
+            log.println("  at " + frame);
+          }
+          try {
+            log.println("Sleeping");
+            Thread.sleep(1000);
+            log.println("Slept 1s, trying again.");
+            return;
+          } catch (InterruptedException ex) {
+            ex.printStackTrace(log);
+          }
+        }
+      }
     }
+    try (var ctx = Context.create()) {
+      log.println("Creating and closing " + ctx);
+    }
+  }
+
+  private static void flushLog(StringWriter buffer) {
+    System.err.println(buffer);
+    buffer.getBuffer().setLength(0);
   }
 }

